File: 3121_Count_the_Number_of_Special_Characters_II/solution.c
int numberOfSpecialChars(char* word) {
    int l[26], f[26];
    memset(l, -1, sizeof(l));
    memset(f, -1, sizeof(f));
    for (int i = 0; word[i]; i++) {
        char c = word[i];
        if (c >= 'a' && c <= 'z') l[c - 'a'] = i;
        else { if (f[c - 'A'] == -1) f[c - 'A'] = i; }
    }
    int r = 0;
    for (int i = 0; i < 26; i++) if (l[i] != -1 && f[i] != -1 && l[i] < f[i]) r++;
    return r;
}

File: 3121_Count_the_Number_of_Special_Characters_II/solution.py
import string

class Solution:
    def numberOfSpecialChars(self, word: str) -> int:
        l, f = {}, {}
        for i, c in enumerate(word):
            if c.islower():
                l[c] = i
            else:
                if c not in f:
                    f[c] = i
        r = 0
        for c in string.ascii_lowercase:
            if c in l and c.upper() in f and l[c] < f[c.upper()]:
                r += 1
        return r
